DOCSP-15527 fundamentals context (#9)
mongodb/docs-golang: source/fundamentals/context.txt

=======
Context
=======

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

> In this guide, you can learn how the MongoDB Go Driver uses the

Author: Agreed -- This is definitely not a guide. 

I do like how it's currently structured. I'm not sure there will be a flow if it were rearranged as you're suggesting.

**context package** from Go's standard library. 

The context package consists of the ``Context`` and ``CancelFunc``
types. The Go Driver uses the ``Context`` type to simplify the
process of creating timeout or cancellation tokens for an operation in
a blocking function. A blocking function is a function that relies

> on an event, such as an input or output, to proceed with its task.

Author: Great catch -- It should definitely be described here! I actually described it in the following sentence but forgot to include it in the description.

Expiration
----------

The driver considers a context expired when an operation exceeds its
timeout or cancellation. It checks the context expiration with the ``Done()``

> function of the ``Context`` type. The ``Done()`` function returns a

Author: Correct, "is canceled" is more accurate -- I initially meant it's time before it's cancelled, but your suggestion better translates it

closed channel and when the function should cancel the work completed.
If you can successfully read from the channel, then the context expired.

> The following sections show when and how the driver checks for

Reviewer: Suggestion: We don't have any diagrams or screenshots here, so I think "describe" is more accurate than "show".

```suggestion
The following sections describe when and how the driver checks for
```

expiration. 

Server Selection
~~~~~~~~~~~~~~~~

When selecting a server for an operation, the driver may block if you

> cannot select a server. 

Author: "You" are telling the driver to select the process. I'll be rewording your suggestion slightly.

The driver loops until it finds a server to use for the operation. After
each iteration, the driver returns a server selection timeout error if
the context expired or the selection process took longer than the
``serverSelectionTimeoutMS`` setting. 

For more information on how the driver selects a server, see the

> :manual:`Server Selection Algorithm Server Manual Page </core/read-preference-mechanics/>`.

Author: Found their page [here](https://github.com/mongodb/docs/blob/master/source/core/read-preference-mechanics.txt). The tag they use is `replica-set-read-preference-behavior`

Connection Checkout
~~~~~~~~~~~~~~~~~~~

After selecting a server, the driver sets a timeout for the
connection checkout process from the server’s connection pool using
the context. If you cannot retrieve a connection before the context
expires, the context returns a timeout error.  

The following example sets the connection timeout to *3* seconds: 

.. code-block:: go
   :emphasize-lines: 2

    opts := options.Client()
    opts.SetConnectTimeout(3 * time.Second)
    client := mongo.Connect(context.TODO(), opts)

Connection Establishment 
~~~~~~~~~~~~~~~~~~~~~~~~

When the driver needs to create a new connection to execute an
operation, the context sets a timeout for the establishment process. The
driver sets the network connection and the TLS handshake to either the
context expiration or connection timeout, whichever is shorter.

The following example sets the connection timeout to *1* second and the
context deadline to *2* seconds. Because the connection timeout is

> shorter, the driver expires after *1* second when establishing a

Reviewer: Question: Is it accurate to say the driver expires? Or is it more like the operation expires?

Author: It's the connection establishment process -- I'll clarify that in my revision.

connection to perform the insert operation.

.. code-block:: go
   :emphasize-lines: 2, 4

    opts := options.Client()
    opts.SetConnectTimeout(1 * time.Second)
    client := mongo.Connect(context.TODO(), opts)
    ctx := context.WithDeadline(2 * time.Second)

    client.Database("<db>").Collection("<collection>").InsertOne(ctx, bson.D{{"x",1}})
    
Socket Read and Write
~~~~~~~~~~~~~~~~~~~~~

When the driver retrieves a connection for the operation, it sets the
socket’s read or write deadline to either the context deadline or socket
timeout, whichever is shorter.

If you cancel the context after the execution of the ``Read()`` or
``Write()`` function but before its deadline, the behavior of the driver

> differs based on version. 

Reviewer: Suggestion: To me, this should be the part that is in the Important admonition: calling out the fact that the behavior differs by version. The next two paragraphs would be normal text.

.. important:: 

>     For versions prior to 1.5.0, the driver won’t detect the context

Reviewer: Change: Use the present tense. The future tense can imply uncertainty, which we don't want. ( [style guide reference](https://docs.mongodb.com/meta/style-guide/writing/use-present-tense/) )

```suggestion
    For versions prior to 1.5.0, the driver doesn't detect the context
```

    cancellation and waits for the ``Read()`` or ``Write()`` function to
    return. 

    In versions 1.5.0 and later, the driver generates a separate
    goroutine to listen for context cancellation when the ``Read()`` or
    ``Write()`` function is in progress. If the goroutine detects a
    cancellation, it closes the connection. The pending ``Read()`` or

>     ``Write()`` function returns with an error that is overwritten to
>     the ``context.Canceled`` property by the driver. 

Reviewer: Change: Avoid passive voice. ( [style guide reference](https://docs.mongodb.com/meta/style-guide/writing/use-active-voice/) )

```suggestion
    ``Write()`` function returns an error that the driver overwrites to
    the ``context.Canceled`` property. 
```

Additional Information
----------------------

> For more information about the context package, see the `official go
> documentation on the context package <https://pkg.go.dev/context>`__.

Reviewer: Ah! Here's the link. I think its ok to have it down here too :)

Change: I think Go is capitalized, or at least it is on their docs site. And we don't need to tell them that its on the context package in the link text, since you introduce it in the first half of the sentence.

```suggestion
For more information about the context package, see the `Go
documentation <https://pkg.go.dev/context>`__.
```

After selecting a server, the driver sets a timeout for the
connection checkout process from the server’s connection pool using
the context. The context returns a timeout error if you cannot retrieve
a connection before it expires.  

The following example sets the connection timeout to *3* seconds: 

.. code-block:: go
   :emphasize-lines: 2

    opts := options.Client()
    opts.SetConnectTimeout(3 * time.Second)
    client := mongo.Connect(context.TODO(), opts)

Connection Establishment 
~~~~~~~~~~~~~~~~~~~~~~~~

When the driver needs to create a new connection to execute an
operation, the context sets a timeout for the establishment process. The
driver sets the timeout to either the context expiration or connection
timeout, whichever is shorter. 

The following example sets the connection timeout to *1* second and the
context deadline to *2* seconds. Because the connection timeout is
shorter, the establishment process expires after *1* second.

.. code-block:: go
   :emphasize-lines: 2, 4

    opts := options.Client()
    opts.SetConnectTimeout(1 * time.Second)
    client := mongo.Connect(context.TODO(), opts)
    ctx := context.WithDeadline(2 * time.Second)

>     client.Database("<db>").Collection("<collection>").InsertOne(ctx, bson.D{{"x",1}})

Reviewer 2: Correct code to:
```go
...
client := mongo.Connect(context.TODO(), opts)

ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
defer cancel()
client.Database("<db>").Collection("<collection>").InsertOne(ctx, bson.D{{"x",1}})
```

    
Socket Read and Write
~~~~~~~~~~~~~~~~~~~~~

When the driver retrieves a connection for an operation, it sets the
socket’s read or write deadline to either the context deadline or socket
timeout, whichever is shorter.

.. important:: 

    If you cancel the context after the execution of the ``Read()`` or
    ``Write()`` function but before its deadline, the behavior of the driver
    differs based on version. 

In versions prior to 1.5.0, the driver doesn't detect the context
cancellation and waits for the ``Read()`` or ``Write()`` function to
return. 

In versions 1.5.0 and later, the driver generates a separate
goroutine to listen for context cancellation when the ``Read()`` or
``Write()`` function is in progress. If the goroutine detects a
cancellation, it closes the connection. The pending ``Read()`` or
``Write()`` function returns an error which the driver overwrites to

> the ``context.Canceled`` property. 

Reviewer 2: Should be
> overwrites with the `context.Canceled` error.

=======
Context
=======

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

The MongoDB Go Driver uses the **context package** from Go's standard
library. 

The context package consists of the ``Context`` and ``CancelFunc``

> types. The Go Driver uses the both types to simplify the process of

Reviewer: Change: Typo.
```suggestion
types. The Go Driver uses both types to simplify the process of
```

function. A blocking function is a function that relies on an event,
such as an input or output, to proceed with its task.

Expiration
----------

The driver considers a context expired when an operation exceeds its
timeout or is canceled. The driver checks the context expiration with
the ``Done()`` function of the ``Context`` type. 

The context is not expired if the function returns a closed channel and
a timestamp that indicates when the function will cancel the completed
work. The context is expired if the function returns an open channel
that you can read from. 

The following sections describe when and how the driver checks for
expiration. 

Server Selection
~~~~~~~~~~~~~~~~

The driver might block you if it can't select a server for an operation.

In this scenario, the driver loops until it finds a server to use for the
operation. After each iteration, the driver returns a server selection
timeout error if the context expired or the selection process took
longer than the ``serverSelectionTimeoutMS`` setting. 

For more information on how the driver selects a server, see the
:ref:`Server Selection Algorithm Server Manual Page

> <replica-set-read-preference-behavior>`. 

Reviewer: Comment: The reason this ref didn't work is because this repository is missing an intersphinx file that tells the autobuilder where to look for references. For example [here's](https://github.com/mongodb/docs-node/blob/master/config/intersphinx.yaml#L6) the one in the Node driver docs.
@terakilobyte Can probably help set that up. :)

Once it's set up, you won't need to include the link text here -- the title will automatically generate from the referenced page.
```
For more informatino on how the driver selects a server, see :ref:`<replica-set-read-preference-bheavior`.
```
That doesn't necessarily need to hold up this review though.

Connection Checkout
~~~~~~~~~~~~~~~~~~~

After selecting a server, the driver sets a timeout for the
connection checkout process from the server’s connection pool using

> context. The context returns a timeout error if you cannot retrieve a

Reviewer: Suggestion: For consistency, should this be "the context"?
```suggestion
the context. The context returns a timeout error if you cannot retrieve a
```

~~~~~~~~~~~~~~~~~~~~~~~~

When the driver needs to create a new connection to execute an
operation, the context sets a timeout for the establishment process. The
driver sets the timeout to either the context expiration or connection
timeout, whichever is shorter. 

The following example sets the connection timeout to *1* second and the
context deadline to *2* seconds. Because the connection timeout is
shorter, the establishment process expires after *1* second.

.. code-block:: go
   :emphasize-lines: 2, 4

    opts := options.Client()
    opts.SetConnectTimeout(1 * time.Second)
    client := mongo.Connect(context.TODO(), opts)
    ctx := context.WithDeadline(2 * time.Second)

    client.Database("<db>").Collection("<collection>").InsertOne(ctx, bson.D{{"x",1}})
    
Socket Read and Write
~~~~~~~~~~~~~~~~~~~~~

When the driver retrieves a connection for an operation, it sets the
socket’s read or write deadline to either the context deadline or socket
timeout, whichever is shorter.

.. important:: 

    If you cancel the context after the execution of the ``Read()`` or
    ``Write()`` function but before its deadline, the behavior of the driver
    differs based on version. 

> For versions prior to 1.5.0, the driver doesn't detect the context

Reviewer: Suggestion: Let's use parallel syntax in these two paragraphs to make the comparison extra clear.

```suggestion
In versions prior to 1.5.0, the driver doesn't detect the context
```

cancellation and waits for the ``Read()`` or ``Write()`` function to
return. 

In versions 1.5.0 and later, the driver generates a separate
goroutine to listen for context cancellation when the ``Read()`` or
``Write()`` function is in progress. If the goroutine detects a
cancellation, it closes the connection. The pending ``Read()`` or

> ``Write()`` function returns an error the driver overwrites to

Reviewer: Change: Typo.
```suggestion
``Write()`` function returns an error which the driver overwrites to
```
